### PR TITLE
Optimize allocation paths in RawVec

### PR DESCRIPTION
**Pull request by alexcrichton**
Since the `Alloc` trait was introduced (https://github.com/rust-lang/rust/pull/42313) and it was integrated everywhere (https://github.com/rust-lang/rust/pull/42727) there's been some slowdowns and regressions that have slipped through. The intention of this PR is to try to tackle at least some of them, but they've been very difficult to quantify up to this point so it probably doesn't solve everything.

This PR primarily targets the `RawVec` type, specifically the `double` function. The codegen for this function is now much closer to what it was before #42313 landed as many runtime checks have been elided.

### CODE REVIEW

**Comment by alexcrichton** [comment]
I'd like to nominate this for beta eventually, but I'd prefer to wait and see the impact on perf.rust-lang.org before deciding to do so.

**Comment by kennytm** [comment]
"fatal runtime error: unsupported allocator request"

```
[00:55:55]      Running build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/collectionstests-9d97829c4a885d0c
[00:55:55] 
[00:55:55] running 461 tests
[00:55:55] thread '<unnamed>' panicked at 'explicit panic', /checkout/src/liballoc/../liballoc/tests/slice.rs:1194:16
[00:55:55] fatal runtime error: unsupported allocator request
[00:55:55] error: An unknown error occurred
```

File: src/liballoc/raw_vec.rs [hunk 1]
```diff
@@ -8,14 +8,13 @@
...
 
-use allocator::{Alloc, Layout};
-use core::ptr::{self, Unique};
+use core::cmp;
 use core::mem;
+use core::ops::Drop;
+use core::ptr::{self, Unique};
 use core::slice;
-use heap::Heap;
+use heap::{Alloc, Layout, Heap, AllocErr};
```
Comment (kennytm) [inline src/liballoc/raw_vec.rs]: Cannot build `tidy` due to unused import.

```rust
[00:02:33]    Compiling alloc v0.0.0 (file:///checkout/src/liballoc)
[00:02:36] error: unused import: `AllocErr`
[00:02:36]   --> /checkout/src/liballoc/raw_vec.rs:16:33
[00:02:36]    |
[00:02:36] 16 | use heap::{Alloc, Layout, Heap, AllocErr};
[00:02:36]    |                                 ^^^^^^^^
[00:02:36]    |
[00:02:36] note: lint level defined here
[00:02:36]   --> /checkout/src/liballoc/lib.rs:77:9
[00:02:36]    |
[00:02:36] 77 | #![deny(warnings)]
[00:02:36]    |         ^^^^^^^^
[00:02:36]    = note: #[deny(unused_imports)] implied by #[deny(warnings)]
[00:02:36] 
[00:02:37] error: aborting due to previous error
[00:02:37] 
[00:02:37] error: Could not compile `alloc`.
```

**Comment by alexcrichton** [comment]
Green!

**Pull request merged**